### PR TITLE
WIP typing router state from daniel work

### PR DESCRIPTION
**Pull request by panaC**

### CODE REVIEW

File: src/renderer/components/catalog/GridTagButton.tsx [hunk 1]
```diff
@@ -19,10 +21,10 @@ interface IBaseProps {
 // ReturnType<typeof mapStateToProps>
 // ReturnType<typeof mapDispatchToProps>
 // tslint:disable-next-line: no-empty-interface
-interface IProps extends IBaseProps, RouteComponentProps {
+interface IProps extends IBaseProps, ReturnType<typeof mapStateToProps> {
 }
 
-class GridTagButton extends React.Component<IProps, undefined> {
+class GridTagButton extends React.Component<IProps> {
```
Comment (danielweck) [inline src/renderer/components/catalog/GridTagButton.tsx]: Can we please continue to use the exact same consistent notation for `extends React.Component<PROPS, STATE>`, which is to use `{}` for unspecified PROPS (not `any`), and `undefined` for unspecified STATE (not `null` or empty).

File: src/renderer/components/catalog/Catalog.tsx [hunk 2]
```diff
@@ -95,11 +94,12 @@ const mapStateToProps = (state: RootState) => ({
         "publication/updateTags",
         "reader/setLastReadingLocation",
     ]),
+    location: state.router.location,
```
Comment (danielweck) [inline src/renderer/components/catalog/Catalog.tsx]: Why use `mapStateToProps` instead of `withRouter()` HOC + `RouteComponentProps`? Is this a personal preference, or is there a technical reason?

File: src/renderer/redux/reducers/index.ts [hunk 3]
```diff
@@ -13,25 +13,28 @@ import { importReducer } from "readium-desktop/common/redux/reducers/import";
...
+    combineReducers({
+        i18n: i18nReducer,
+        reader: readerReducer,
+        opds: opdsReducer,
+        win: winReducer,
+        net: netReducer,
+        update: updateReducer,
+        api: apiReducer,
+        dialog: dialogReducer,
+        router: connectRouter<IRouterLocationState>(history),
```
Comment (danielweck) [inline src/renderer/redux/reducers/index.ts]: Great :)

File: src/renderer/components/catalog/GridTagButton.tsx (hunk 1 above)
Comment (panaC) [reply src/renderer/components/catalog/GridTagButton.tsx]: if you want, but it's not really important, it is `{}` by default

Comment (panaC) [reply src/renderer/components/catalog/GridTagButton.tsx]: of course not `any`or `null`

File: src/renderer/components/catalog/SearchForm.tsx [hunk 4]
```diff
@@ -62,18 +62,20 @@ class Search extends React.Component<IProps, undefined> {
...
         const value = this.inputRef?.current?.value;
         if (!value) {
-            this.props.redirect({
-                pathname: "/library/search/all",
-                search: undefined,
+            this.props.history.push({
+                ...this.props.location,
+                ...{
+                    pathname: "/library/search/all",
+                },
```
Comment (danielweck) [inline src/renderer/components/catalog/SearchForm.tsx]: In my tests, and according to the React Router documentation, the `history` API does not handle `state`, thus why I used the `<Redirect/>` React Component (as a HOC, but we can also use it directly inside each implicated component class).
I will have to run tests again with this PR again to find the edge cases where the state (`displayType`) is not preserved across navigations, for example inside OPDS (search and browsing), inside local bookshelf (search and browsing), as well as when navigating across OPDS/local bookshelf.

File: src/renderer/components/catalog/GridTagButton.tsx (hunk 1 above)
Comment (danielweck) [reply src/renderer/components/catalog/GridTagButton.tsx]: It is just for code style consistency, search and replace, etc.
